### PR TITLE
[OKTA-408185] Restart login flow on cancel

### PR DESCRIPTION
**Pull request by denysoblohin-okta**
## Description:
Issue: User can't login after clicking 'Back to sign in' (see video)
JS error: 
```js
TypeError: Cannot read property 'state' of null
```
on this line:
https://github.com/okta/okta-signin-widget/blob/8d27ecb7f46fb1ea11ad49cf8538b661de546239/src/v2/client/interactionCodeFlow.js#L23

Source of error: After click on 'Back to sign in' idx transaction is canceled and transaction meta is cleared.
Fix method is to restart login flow after clear-up

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [x] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video of error this PR is fixing:

https://user-images.githubusercontent.com/72614880/124826104-1fa8e300-df7d-11eb-9c52-750f36ca0a3e.mov

Screen Shot 2021-06-28 at 10 21 27 PM

### Issue:

- [OKTA-408185](https://oktainc.atlassian.net/browse/OKTA-408185)

### CODE REVIEW

File: src/v2/controllers/FormController.js
Comment (shuowu) [inline src/v2/controllers/FormController.js:159]: How about emit an event and let the `BaseLoginRouter` listen on it? in this way, we can centralize the reset/init logic in one place.

File: src/models/Settings.js
Comment (shuowu) [inline src/models/Settings.js:537]: When the transaction is canceled, a new one should be started when back to the login screen. No preserve logic is needed here.

Comment (denysoblohin-okta) [reply src/models/Settings.js:537]: After starting login flow these 3 options are unset:
https://github.com/okta/okta-signin-widget/blob/8d27ecb7f46fb1ea11ad49cf8538b661de546239/src/v2/BaseLoginRouter.js#L216-L218
This will prevent further interaction:
https://github.com/okta/okta-signin-widget/blob/8d27ecb7f46fb1ea11ad49cf8538b661de546239/src/v2/client/startLoginFlow.js#L37-L40

Comment (shuowu) [reply src/models/Settings.js:537]: `stateToken` should be unset in a new transaction. `proxyIdxResponse` is probably used for debugging only. I only setting you want to preserve is `useInteractionCodeFlow`, which can be set to true programmatically when an event is triggered. We can probably name it as `interactionCanceled`.

File: src/v2/controllers/FormController.js
Comment (pradeepdewda-okta) [reply src/v2/controllers/FormController.js:159]: We should not initiate login flow from FormController..

Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:164]: @shuowu @pradeepdewda-okta Please review changes

**Comment by pradeepdewda-okta** [comment]
Can you add a video? Also add debugger screen.

Can you verify that we are cleaning up `stateToken` from sessionStorage properly.. You can check that in debugger

File: src/v2/controllers/FormController.js
Comment (pradeepdewda-okta) [reply src/v2/controllers/FormController.js:164]: I think instead, you should trigger BaseLoginRouter's `render` function instead that starts the loginflow and handle cleanup..

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:189]: instead of creating this new handler, it should use existing `startLoginFlow` of `render`

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:192]: this creates duplication and alienate `interact` from main flow, hard to maintain widget state properly in case of errors..

**Comment by denysoblohin-okta** [comment]
@pradeepdewda-okta 
I've created `startLoginFlow` method in login router to reuse in `render` and `interactionCanceled` handler.
In my understanding `interact` is necessary to start a new login flow after cancel, because `interactionHandle` has been cleared on `cancel`. So not sure how to eliminate this alienation.

Added video and screenshot of error.

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:83]: does it not work if you do `this.render` instead of `this.startLoginFlow`

**Comment by pradeepdewda-okta** [comment]
Couldn't understand the video.. do we expect to see error here?  that doesn't seem right

**Comment by denysoblohin-okta** [comment]
@pradeepdewda-okta 
Video of error:

https://user-images.githubusercontent.com/72614880/124838218-10329580-df8f-11eb-920d-5d56df1120c2.mov


Video of fix:

https://user-images.githubusercontent.com/72614880/124837987-99959800-df8e-11eb-84fc-3959fa62278f.mov

We do not expect error. This PR should fix the error

File: src/v2/BaseLoginRouter.js
Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js:83]: No, this would produce error `TypeError: Controller is not a constructor` in this line:
https://github.com/okta/okta-signin-widget/blob/8d27ecb7f46fb1ea11ad49cf8538b661de546239/src/v2/BaseLoginRouter.js#L247

Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js]: Changed `render` method for using it for re-render current controller (without parameters)

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:164]: Done

File: src/v2/BaseLoginRouter.js
Comment (aarongranick-okta) [inline src/v2/BaseLoginRouter.js:256]: In what case do we not have `Controller` ? In that case, won't the next line
```
this.controller.render()
```
throw an exception? (no property render on undefined)

File: test/testcafe/spec/SessionStorage_spec.js
Comment (aarongranick-okta) [inline test/testcafe/spec/SessionStorage_spec.js:281]: It looks like this may have been copied from `EnglishLeaks_spec`. This config appears to be incorrect / misleading.

- `codeVerifier` is not a widget option. (It may exist in transaction meta)
- `codeChallenge` option will put the widget into a different mode (remediation, meant for server-side web apps). In this mode the widget will not retrieve tokens. Although the token endpoints are mocked for success, I believe this part of the flow is not being tested.
- `codeChallengeMethod` default value is `S256`. Recommend not setting this option in the test

It would be great if you could clean up the config here and in `EnglishLeaks`:
- remove `codeVerifier`, `codeChallenge` and `codeChallengeMethod` options. This should not break the tests. If, for some reason, removing these options does cause any change, we should test both cases. (with/without codeChallenge)

File: src/v2/controllers/FormController.js
Comment (aarongranick-okta) [inline src/v2/controllers/FormController.js:140]: Why are we using presence of interactionHandle in storage as "source of truth" instead of settings.get('useInteractionFlow')? How could we have transaction storage containing an interactionHandle if our config does not have `useInteractionCodeFlow`?

Because the widget is currently accepting an option, `interactionHandle` which is accepted instead of `useInteractionCodeFlow`:
https://github.com/okta/okta-signin-widget/blob/master/src/v2/client/startLoginFlow.js#L37
https://github.com/okta/okta-signin-widget/blob/master/src/models/Settings.js#L247

There is a task to remove this option, as it has been deemed an anti-pattern: OKTA-405474
Please add a code comment here
```
// TODO: use settings as SoT for `useInteractionCodeFlow` OKTA-405474
```

File: src/v2/controllers/FormController.js
Comment (aarongranick-okta) [inline src/v2/controllers/FormController.js:167]: Please add a code comment here
```
// TODO: remove this line after OKTA-405474
```

File: src/v2/BaseLoginRouter.js
Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js:256]: `Controller` will be undefined only when `render` is called on `interactionCanceled` handler:
https://github.com/okta/okta-signin-widget/blob/d3b02d7a0a3de1befdd71eddd5948d9fc64d8c95/src/v2/BaseLoginRouter.js#L83
In this case `this.controller` should be already defined.

Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:256]: It is because when you trigger `interactionCanceled` the `controller` context should be passed and then you can call render on that context?

File: src/v2/controllers/FormController.js
Comment (pradeepdewda-okta) [inline src/v2/controllers/FormController.js:166]: does this need to be taken care in `handleSaveForm` as well ?

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:140]: We can't rely on `useInteractionFlow` because it's cleared on this line:
https://github.com/okta/okta-signin-widget/blob/9d9dd22e24a2938548ecfbeb3b69f16a8bfb20fd/src/v2/BaseLoginRouter.js#L219

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:256]: `Back to sign in` is basically a redirect to login page and also clears any stateToken stored in sessionStorage. Can't we do something similar for `interactionCodeFlow` ? like just delete any existing token and do a page refresh to restart the login flow

File: src/v2/controllers/FormController.js
Comment (aarongranick-okta) [reply src/v2/controllers/FormController.js:140]: That's odd, I wonder why it is being unset. Will be investigated along with OKTA-405474.

Can you also tag these lines with 
```
// TODO: do not unset  `useInteractionCodeFlow` OKTA-405474
```

File: test/testcafe/spec/SessionStorage_spec.js
Comment (denysoblohin-okta) [reply test/testcafe/spec/SessionStorage_spec.js:281]: Done. It has not broken tests

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:140]: Added

File: src/v2/BaseLoginRouter.js
Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js:256]: Reverted changes to `render` method. Pass controller as context to `interactionCanceled` event

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:166]: In my understanding, no. As we need to restart login flow only on cancel, which is triggered from here
https://github.com/okta/okta-signin-widget/blob/194575de499cb4ce91453362a517623256dbab87/src/v2/view-builder/components/Link.js#L57
and handled in `handleInvokeAction`

Comment (pradeepdewda-okta) [reply src/v2/controllers/FormController.js:166]: Why only on `cancel` ? can there be token/code expiration ? what if enduser do a hard refresh on this page when token is expired ?

Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js]: If token is expired, user will see screenshot 1. Click on 'Back to sign-in' will restart login flow correctly.
Screenshot 2021-07-12 at 22 19 17

If user will do a hard refresh with expired token, user will see screenshot 2. Click on  'Back to sign-in' will do nothing as it's just goto link. I fixed this in last commit by setting `actionPath: cancel` and removing `'href': settings?.get('baseUrl'),` 
Screenshot 2021-07-12 at 22 19 46

WDYT?

File: src/v2/view-builder/utils/LinksUtil.js
Comment (pradeepdewda-okta) [inline src/v2/view-builder/utils/LinksUtil.js:128]: why this ? this can't be commented.. It will break `Back to sign in` link for various flows.

Comment (denysoblohin-okta) [reply src/v2/view-builder/utils/LinksUtil.js:128]: I've removed `href` and added `actionPath` because they can't work together according to this code:
https://github.com/okta/okta-signin-widget/blob/56930aa682bf8e0aee294b5a1c8588c5b0d4c6a8/src/v2/view-builder/components/Link.js#L41-L57

Comment (pradeepdewda-okta) [reply src/v2/view-builder/utils/LinksUtil.js:128]: @denysoblohin-okta there is already a version for that `getSignOutLink`

Comment (denysoblohin-okta) [reply src/v2/view-builder/utils/LinksUtil.js:128]: Another possible solution I see is to catch specific `idx.session.expired` error and only in this case trigger `cancel` action on click on 'Back to sign-in' , in other cases use href

File: src/v2/controllers/FormController.js
Comment (jmelberg-okta) [inline src/v2/controllers/FormController.js:165]: This is going to create problems for us, as it appears we executing this handler after the resolved promise from the [`handleIdxResponseSuccess`](https://github.com/okta/okta-signin-widget/blob/55a5d9cad217477abe56de8ab5dd828546e54888/src/v2/BaseLoginRouter.js#L85) method.

We already check for the `actionPath` above (see line 141) - why can't this logic sit there?

File: test/testcafe/spec/SessionStorage_spec.js
Comment (jmelberg-okta) [inline test/testcafe/spec/SessionStorage_spec.js:159]: This test is asserting the `stateHandle` is empty, so we're breaking existing behavior with this change.

What do we expect this value to be now?

File: test/testcafe/spec/SessionStorage_spec.js
Comment (jmelberg-okta) [inline test/testcafe/spec/SessionStorage_spec.js:320]: Instead of checking for a prompt, we can get more value from parsing the console logs to validate the correct [response](https://github.com/okta/okta-signin-widget/blob/17682c785c97f00a6e53911c455c61841b2798ae/playground/main.js#L62).

File: src/v2/BaseLoginRouter.js
Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js:227]: Adding this comment was requested by @aarongranick-okta : https://github.com/okta/okta-signin-widget/pull/2090#discussion_r667134521
Option `useInteractionCodeFlow ` will be removed in future.

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:138]: Because `useInteractionCodeFlow` is unset after starting login flow here:
https://github.com/okta/okta-signin-widget/blob/56930aa682bf8e0aee294b5a1c8588c5b0d4c6a8/src/v2/BaseLoginRouter.js#L218
The only way I see to "compute" this property after being deleted is to check `interactionHandle` in transaction meta.
And to load transaction meta we need to use auth-js's manager: 
https://github.com/okta/okta-signin-widget/blob/56930aa682bf8e0aee294b5a1c8588c5b0d4c6a8/src/v2/client/interactionCodeFlow.js#L22

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:165]: I assumed that first we need to cancel current login flow, and then after `/cancel` response we can  start a new one.

Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:165]: But this logic can be moved to line 141

File: test/testcafe/spec/SessionStorage_spec.js
Comment (denysoblohin-okta) [reply test/testcafe/spec/SessionStorage_spec.js:159]: Because of starting new login flow after cancel (the purpose of this PR), state handle should be a new one.

File: src/v2/controllers/FormController.js
Comment (haishengwu-okta) [inline src/v2/controllers/FormController.js:174]: Is there real use case for this code branch? 
IIRC, There is no way that widget triggers `cancel` action but `actions[cancel]` is not function.

File: src/v2/controllers/FormController.js
Comment (haishengwu-okta) [inline src/v2/controllers/FormController.js:161]: I wonder if you shall suppose to change here instead of adding another `.then` ?
```
.then(function () {
  if (action === 'cancel' AND useInteractionFlow) {
    this.options.appState.trigger('interactionCanceled');
  } else {
     this.handleIdxSuccess();
  }
});
```

File: src/v2/controllers/FormController.js
Comment (haishengwu-okta) [inline src/v2/controllers/FormController.js:171]: why need to pass `this.constructor` ?

File: src/v2/controllers/FormController.js
Comment (haishengwu-okta) [inline src/v2/controllers/FormController.js:139]: Could create a help function in `client/transactionMeta.js`.

File: test/testcafe/spec/SessionStorage_spec.js
Comment (denysoblohin-okta) [reply test/testcafe/spec/SessionStorage_spec.js:320]: It would be great, but TestCafe can't return object used in `console.log`, it returns string `[object Object]`
There is open issue for this:
https://github.com/DevExpress/testcafe/issues/5425

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:157]: Moved this logic instead of `this.handleIdxSuccess`  as suggested by @haishengwu-okta 
https://github.com/okta/okta-signin-widget/pull/2090#discussion_r668471099

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js]: Done

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js:161]: Because this will be passed as first parameter (`Controller`) to `BaseLoginRouter.render()` 
https://github.com/okta/okta-signin-widget/blob/1bca146103174f2993c2412b3fe7bc7a3e7fd7a8/src/v2/BaseLoginRouter.js#L83
https://github.com/okta/okta-signin-widget/blob/1bca146103174f2993c2412b3fe7bc7a3e7fd7a8/src/v2/BaseLoginRouter.js#L189

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js]: Done

File: src/v2/client/transactionMeta.js
Comment (jmelberg-okta) [inline src/v2/client/transactionMeta.js:102]: Can we reuse the [`getTransactionMeta`](https://github.com/okta/okta-signin-widget/blob/c07ce5a87879a4bf98faf40cf8239eda8f4d86e9/src/v2/client/transactionMeta.js#L21) method here?

File: test/testcafe/spec/SessionStorage_spec.js
Comment (jmelberg-okta) [reply test/testcafe/spec/SessionStorage_spec.js:159]: Right, my comment was mostly focused on ensuring the assertions match the description.

Let's make sure we update the test description and have an assertion for the new `stateHandle` we'd expect.

File: test/testcafe/spec/SessionStorage_spec.js
Comment (jmelberg-okta) [reply test/testcafe/spec/SessionStorage_spec.js:320]: We should still be able to compare object equality though, as we're doing using the `checkConsoleMessages` helper. Here is an example where we're [checking for context](https://github.com/okta/okta-signin-widget/blob/d0f55fee688fe6c79620cf7e37e7a53b236311e4/test/testcafe/spec/IdentifyRegistrationWithCaptcha_spec.js#L49-L52).

**Review by jmelberg-okta — COMMENTED** [review]
Overall these changes appear good to me. Let's wait for final approval from Aaron before proceeding with this change.

File: package.json
Comment (haishengwu-okta) [inline package.json:43]: assume this is not intentional?

Comment (denysoblohin-okta) [reply package.json]: Yes, reverted.
(I changed this temporary to see Testcafe errors in Travis)

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js]: Will move fix for expired session to another PR

File: src/v2/view-builder/utils/LinksUtil.js
Comment (denysoblohin-okta) [reply src/v2/view-builder/utils/LinksUtil.js]: Reverted these changes. Will move fix for expired session to another PR

File: src/v2/controllers/FormController.js
Comment (denysoblohin-okta) [reply src/v2/controllers/FormController.js]: Reverted these changes. Will move fix for expired session to another PR

File: src/v2/client/transactionMeta.js
Comment (denysoblohin-okta) [reply src/v2/client/transactionMeta.js:102]: Yes.. but I think it's better to not use `getTransactionMeta` because it's async and will create new transaction meta if one does not exist

File: test/testcafe/spec/SessionStorage_spec.js
Comment (denysoblohin-okta) [reply test/testcafe/spec/SessionStorage_spec.js]: Changed.
See https://github.com/okta/okta-signin-widget/pull/2090/files#diff-d5619ca161ed5c74752d62d68598cc9038e1a588eee3755c9f3a103ed3ad43e0R317

**Review by aarongranick-okta — APPROVED** [review]

**Review by pradeepdewda-okta — CHANGES_REQUESTED** [review]
Beta version of this change needs to pass bacon first..

File: src/v2/controllers/FormController.js
Comment (pradeepdewda-okta) [reply src/v2/controllers/FormController.js:158]: Let's setup a meeting to discuss this change.. I am not comfortable with this one-off chaining here for interaction flow.. why it needs a special treatment.

File: test/testcafe/spec/SessionStorage_spec.js
Comment (pradeepdewda-okta) [inline test/testcafe/spec/SessionStorage_spec.js:267]: typo: succesfully

File: test/testcafe/spec/SessionStorage_spec.js
Comment (pradeepdewda-okta) [inline test/testcafe/spec/SessionStorage_spec.js:7]: remove `.json`

File: playground/main.js
Comment (jmelberg-okta) [inline playground/main.js:58]: I don't think we need this extra notice for success, as it'll complicate the assertion in our testcafe helpers.

I suggest keeping this as it was before, and letting the `JSON.parse(log[i])` logic handle parsing it so we can check the content through `checkConsoleMessages`

Comment (pradeepdewda-okta) [reply playground/main.js:58]: +1

**Comment by pradeepdewda-okta** [comment]
@denysoblohin-okta when you click on `Back to sign in` what does `interact` request use as `interactionHandle` ? is it using a stale token from previous request.. basically upon `cancel` everything should be wiped out..

File: playground/main.js
Comment (denysoblohin-okta) [reply playground/main.js]: Changed to parsing JSON

**Comment by denysoblohin-okta** [comment]
@pradeepdewda-okta `/interact` will return new interaction handle

**Comment by pradeepdewda-okta** [comment]
> @pradeepdewda-okta `/interact` will return new interaction handle. `code_challenge ` and `state` in request are new

Was it null when it does an `interact` without your changes.. ?

**Comment by denysoblohin-okta** [comment]
@pradeepdewda-okta Without my changes after `cancel` there is no `interact` request

File: src/v2/controllers/FormController.js
Comment (pradeepdewda-okta) [reply src/v2/controllers/FormController.js:158]: @denysoblohin-okta will create jira to remove exclusive handling of cancel action for `interaction code flow` for devex team to resolve

**Review by pradeepdewda-okta — APPROVED** [review]
As part of OKTA-405474, TODO's should be fixed

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:229]: can you explain why we need this in comments, instead what it is doing?

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:225]: if we are doing the `render` in finally why we need it here as well?

Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js:225]: The problem I'm trying to solve is the recursive call of `this.render` in then/catch.
If we not unset `useInteractionCodeFlow `, `startLoginFlow` will be called again and result in infinite loop.
`useInteractionCodeFlow ` setting was loop-breaker.

Instead of this, I moved actual render logic into `_render` method and call it in `finally`.

File: src/v2/BaseLoginRouter.js
Comment (denysoblohin-okta) [reply src/v2/BaseLoginRouter.js:229]: I just moved these lines from `render` method into `_render`

Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:229]: i mean why it needs to be in `render`

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [reply src/v2/BaseLoginRouter.js:225]: there seems to be some flaw with this flow, i am not super confident with this PR.. @aarongranick-okta you should take a look on why `useInteractionCodeFlow` unset is a loop breaker. It should not have that dependency.

Comment (aarongranick-okta) [reply src/v2/BaseLoginRouter.js:232]: It seems that `render` is following a recursive pattern, used here to load languages:
https://github.com/okta/okta-signin-widget/blob/master/src/v2/BaseLoginRouter.js#L206

The lines of code to "unset" stateToken, etc. seem to be related to this function, which is not designed to be called recursively.
https://github.com/okta/okta-signin-widget/blob/master/src/v2/client/startLoginFlow.js#L26

Conceptually we want render to move forward, not in a loop. One way to accomplish this would be to use `async/await`

- await LanguageUtil.loadLanguage()
- await startLoginFlow
- continue with normal render

Although some methods return Q promises, I believe it should work. (Some tests may need minor adjustments).

File: src/v2/BaseLoginRouter.js
Comment (pradeepdewda-okta) [inline src/v2/BaseLoginRouter.js:195]: remove this line

File: test/unit/spec/v2/client/transactionMeta_spec.js
Comment (pradeepdewda-okta) [inline test/unit/spec/v2/client/transactionMeta_spec.js:234]: nit-- use lint to remove unnecessary changes from PR

**Review by aarongranick-okta — APPROVED** [review]
lgtm

**Pull request closed without merging**
